src/Nethermind/Nethermind.Synchronization/Trie/HealingStateTree.cs

```diff
@@ -0,0 +1,92 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Diagnostics;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+using Nethermind.State;
+using Nethermind.State.Snap;
+using Nethermind.Trie;
+using Nethermind.Trie.Pruning;
+
+namespace Nethermind.Synchronization.Trie;
+
+public class HealingStateTree : StateTree
+{
+    private ITrieNodeRecovery<GetTrieNodesRequest>? _recovery;
+
+    [DebuggerStepThrough]
+    public HealingStateTree(ITrieStore? store, ILogManager? logManager)
+        : base(store, logManager)
+    {
+    }
+
+    public void InitializeNetwork(ITrieNodeRecovery<GetTrieNodesRequest> recovery)
+    {
+        _recovery = recovery;
+    }
+
+    public override byte[]? Get(ReadOnlySpan<byte> rawKey, Keccak? rootHash = null)
+    {
+        try
+        {
+            return base.Get(rawKey, rootHash);
+        }
+        catch (MissingTrieNodeException e)
+        {
+            if (Recover(e.GetPathPart(), rootHash ?? RootHash))
+            {
+                return base.Get(rawKey, rootHash);
+            }
+
+            throw;
+        }
+    }
+
+    public override void Set(ReadOnlySpan<byte> rawKey, byte[] value)
+    {
+        try
+        {
+            base.Set(rawKey, value);
+        }
+        catch (MissingTrieNodeException e)
+        {
+            if (Recover(e.GetPathPart(), RootHash))
+            {
+                base.Set(rawKey, value);
+            }
+            else
+            {
+                throw;
+            }
+        }
+    }
+
+    private bool Recover(ReadOnlySpan<byte> pathPart, Keccak rootHash)
+    {
+        if (_recovery?.CanRecover == true)
+        {
+            GetTrieNodesRequest request = new()
+            {
+                RootHash = rootHash,
+                AccountAndStoragePaths = new[]
+                {
+                    new PathGroup
+                    {
+                        Group = new[] { Nibbles.EncodePath(pathPart) }
+                    }
+                }
+            };
+
+            byte[]? rlp = _recovery.Recover(request).GetAwaiter().GetResult();
```

Comment:
😢

Reply:
Better ideas? It works :) And we are already broken...

Reply:
The blocking is on a dedicated thread so won't cause thread starvation; so is fine. Still 😢